`storage`: pass `abort_source` to `maybe_create_contiguous_segment_set()`

WillemKauf opened the pull request:
To better respect potential shutdowns during `segment` recovery at start-up.


## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

dotnwat commented on src/v/storage/segment_set.cc:
```diff
@@ -198,6 +198,9 @@ maybe_create_contiguous_segment_set(segment_set::underlying_t segs) {
                         .get_dirty_offset();
     segment_set::underlying_t ignored_segs;
     for (auto it = std::next(segs.begin()); it != segs.end();) {
+        if (as.abort_requested()) {
+            co_return segment_set(std::move(segs));
+        }
```
this return value is surprising. like for example it might even return segs with some of them erased (`it = segs.erase(it);` below), but it woudl be indistinguishable from when the function reached the bottom and all kind of other things had happened. should we return an indication that abort has occurred?

WillemKauf replied:
Right, I was trying to reason about what made more sense as a return value here for the early abort case - the current segment set or `std::nullopt`? I went with the current segment set to be close in spirit to the usage of the `abort_source` in the related function above this one in the call stack:

https://github.com/redpanda-data/redpanda/blob/eac0c1f3ca8e7cb62fdd84aade440544a03a63eb/src/v/storage/segment_set.cc#L382-L386

I have no strong opinions here, if the `abort_source` is triggered that means we're (ostensibly) shutting down early anyways and this return value won't ever really be used, nor will the side effects of the early return here matter since the process will repeat on next start-up/management of the current `log`.

dotnwat replied:
>  if the abort_source is triggered that means we're (ostensibly) shutting down early anyways and this return value won't ever really be used, nor will the side effects of the early return here matter since the process

as an internal interface i think it's cool to co-design the interfaces with this sort of coupling. in general it's better to design an interface to minimize the assumptions that the rules needs to obey.

anyway, does "return value won't ever really be used" hold in this case?

WillemKauf replied:
The source abort source comes from the `log_manager`, which only triggers `request_abort()` when `log_manager::stop()` is called:

https://github.com/redpanda-data/redpanda/blob/5159ef278364beb27ab3c3f39db68a863cb5b82f/src/v/storage/log_manager.cc#L199-L202

Though, the return value from `recover_segments()` here:

https://github.com/redpanda-data/redpanda/blob/5159ef278364beb27ab3c3f39db68a863cb5b82f/src/v/storage/log_manager.cc#L772-L782

Is used in the call to `make_disk_backed_log()` below.

https://github.com/redpanda-data/redpanda/blob/5159ef278364beb27ab3c3f39db68a863cb5b82f/src/v/storage/log_manager.cc#L784-L791

So when I say "won't ever be used", I suppose I mean that it is expected that this is only triggered during shutdown, and the `disk_log_impl`s will be `close()`'d imminently, though I suppose this may not quite be airtight in guaranteeing "no use" of the returned `segment_set` value.

We could add another abort source check after the call to `recover_segments()` on L772 in `log_manager.cc` but before the `log` is created to instead throw an exception to ensure the return value is _really_ never used in the case of an early shutdown/return.

WDYT?

dotnwat replied:
I'm finding it quite odd that it looks like log recovery can have an abort be triggered, and then seemingly construct a log instance with an incomplete? set of segments depending on which the abort happened? i traced it back to the incoming log_manager::manage, which is even weirder because then does the log go live?

WillemKauf replied:
From `partition_manager`, `log_manager::manage()` is called:

https://github.com/redpanda-data/redpanda/blob/98cf9ebcb8dadce94a7a0575212118ee59d19116/src/v/cluster/partition_manager.cc#L278-L287

Afterwards, `consensus` object and `partition` objects are created, finally `partition::start()` is called:

https://github.com/redpanda-data/redpanda/blob/98cf9ebcb8dadce94a7a0575212118ee59d19116/src/v/cluster/partition_manager.cc#L289-L324

From `partition::start()`, `consensus::start()` is invoked:

https://github.com/redpanda-data/redpanda/blob/98cf9ebcb8dadce94a7a0575212118ee59d19116/src/v/cluster/partition.cc#L462-L471

Which finally calls `log->start()`.

https://github.com/redpanda-data/redpanda/blob/98cf9ebcb8dadce94a7a0575212118ee59d19116/src/v/raft/consensus.cc#L1526

If the recovery is stopped because an abort is triggered, we may as well check for it immediately after trying to recover segments.


```diff
ss::future<ss::shared_ptr<log>> log_manager::do_manage(
  ntp_config cfg,
  raft::group_id group,
  std::vector<model::record_batch_type> translator_batch_types) {
    if (_config.base_dir.empty()) {
        throw std::runtime_error(
          "log_manager:: cannot have empty config.base_dir");
    }

    vassert(
      _logs.find(cfg.ntp()) == _logs.end(), "cannot double register same ntp");

    std::optional<ss::sstring> last_clean_segment;
    auto clean_iobuf = _kvstore.get(
      kvstore::key_space::storage, internal::clean_segment_key(cfg.ntp()));
    if (clean_iobuf) {
        last_clean_segment = serde::from_iobuf<internal::clean_segment_value>(
                               std::move(clean_iobuf.value()))
                               .segment_name;
    }

    co_await maybe_clear_kvstore(cfg);

    with_cache cache_enabled = cfg.cache_enabled();
    auto ntp_sanitizer_cfg = _config.maybe_get_ntp_sanitizer_config(cfg.ntp());

    auto segments = co_await recover_segments(
      partition_path(cfg),
      cfg.is_compacted(),
      [this, cache_enabled] { return create_cache(cache_enabled); },
      _abort_source,
      config::shard_local_cfg().storage_read_buffer_size(),
      config::shard_local_cfg().storage_read_readahead_count(),
      last_clean_segment,
      _resources,
      _feature_table,
      std::move(ntp_sanitizer_cfg));

+   _abort_source.check();

    auto l = storage::make_disk_backed_log(
      std::move(cfg),
      group,
      *this,
      std::move(segments),
      _kvstore,
      _feature_table,
      std::move(translator_batch_types));
```

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--381e08b7-39a6-4020-a0dc-c67ccb51d2ee-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69810">build#69810</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:---------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/69810#01985449-fb32-4659-8a86-530afe388053 | FLAKY         | 20/21    | upstream reliability is '99.57627118644068'. current run reliability is '95.23809523809523'. drift is 4.33818 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

nvartolomei commented:
Realistically, for how long can this hang?

WillemKauf commented:
> Realistically, for how long can this hang?

Probably not long, I would assume on the order of seconds at most, but I thought it was worth plumbing the `abort_source` in (since there is one in the call stack above this function that is used in the same way).

nvartolomei commented:
That's a lot of context to keep in your head to reason about correctness.
Why we don't throw if recovery is aborted? To me that looks like way more
straightforward mechanism and then you don't have to worry at all about
someone using a bogus return value inadvertently.

On Fri, Aug 1, 2025 at 3:27 PM Willem Kaufmann ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/v/storage/segment_set.cc
> <https://github.com/redpanda-data/redpanda/pull/27025#discussion_r2248133799>
> :
>
> > +        if (as.abort_requested()) {
> +            co_return segment_set(std::move(segs));
> +        }
>
> From partition_manager, log_manager::manage() is called:
>
>
> https://github.com/redpanda-data/redpanda/blob/98cf9ebcb8dadce94a7a0575212118ee59d19116/src/v/cluster/partition_manager.cc#L278-L287
>
> Afterwards, consensus object and partition objects are created, finally
> partition::start() is called:
>
>
> https://github.com/redpanda-data/redpanda/blob/98cf9ebcb8dadce94a7a0575212118ee59d19116/src/v/cluster/partition_manager.cc#L289-L324
>
> From partition::start(), consensus::start() is invoked:
>
>
> https://github.com/redpanda-data/redpanda/blob/98cf9ebcb8dadce94a7a0575212118ee59d19116/src/v/cluster/partition.cc#L462-L471
>
> Which finally calls log->start().
>
>
> https://github.com/redpanda-data/redpanda/blob/98cf9ebcb8dadce94a7a0575212118ee59d19116/src/v/raft/consensus.cc#L1526
>
> If the recovery is stopped because an abort is triggered, we may as well
> check for it immediately after trying to recover segments.
>
> ss::future<ss::shared_ptr<log>> log_manager::do_manage(
>   ntp_config cfg,
>   raft::group_id group,
>   std::vector<model::record_batch_type> translator_batch_types) {
>     if (_config.base_dir.empty()) {
>         throw std::runtime_error(
>           "log_manager:: cannot have empty config.base_dir");
>     }
>
>     vassert(
>       _logs.find(cfg.ntp()) == _logs.end(), "cannot double register same ntp");
>
>     std::optional<ss::sstring> last_clean_segment;
>     auto clean_iobuf = _kvstore.get(
>       kvstore::key_space::storage, internal::clean_segment_key(cfg.ntp()));
>     if (clean_iobuf) {
>         last_clean_segment = serde::from_iobuf<internal::clean_segment_value>(
>                                std::move(clean_iobuf.value()))
>                                .segment_name;
>     }
>
>     co_await maybe_clear_kvstore(cfg);
>
>     with_cache cache_enabled = cfg.cache_enabled();
>     auto ntp_sanitizer_cfg = _config.maybe_get_ntp_sanitizer_config(cfg.ntp());
>
>     auto segments = co_await recover_segments(
>       partition_path(cfg),
>       cfg.is_compacted(),
>       [this, cache_enabled] { return create_cache(cache_enabled); },
>       _abort_source,
>       config::shard_local_cfg().storage_read_buffer_size(),
>       config::shard_local_cfg().storage_read_readahead_count(),
>       last_clean_segment,
>       _resources,
>       _feature_table,
>       std::move(ntp_sanitizer_cfg));
> +   _abort_source.check();
>
>     auto l = storage::make_disk_backed_log(
>       std::move(cfg),
>       group,
>       *this,
>       std::move(segments),
>       _kvstore,
>       _feature_table,
>       std::move(translator_batch_types));
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/redpanda-data/redpanda/pull/27025#discussion_r2248133799>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEETWPATX74OFX2JLQ25433LN2OTAVCNFSM6AAAAACCSR4WCGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTANZZGQ3DMMRRHE>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

WillemKauf committed:
`storage`: pass `abort_source` to `maybe_create_contiguous_segment_set()`

To better respect potential shutdowns during `segment` recovery at start-up.

WillemKauf committed:
`storage`: call `abort_source::check()` during recovery

Instead of returning a potentially truncated set of `segment`s during
recovery in the `storage` layer, call `abort_source::check()` to instead
throw an exception back up the stack to prevent use of a potentially
rubbish `segment_set`.

WillemKauf commented:
Force push to:
* Call `abort_source::check()` instead of returning from functions during `segment_set` recovery to prevent use of truncated sets elsewhere in the code.

WillemKauf commented:
> That's a lot of context to keep in your head to reason about correctness.
Why we don't throw if recovery is aborted? To me that looks like way more
straightforward mechanism and then you don't have to worry at all about
someone using a bogus return value inadvertently.

Thanks, I do agree. I was trying to find a way to maintain status quo behavior, but I agree, `segment_set` recovery should instead behave this way.

nvartolomei commented:
One thing I wonder about is whether the previous behavior was there to
support making incremental progress even if the server is shutdown. But I
doubt it. But if that was the case, previously the code was somewhat
correct for such a use-case but the new return you were adding I think
would have broken it.

On Fri, Aug 1, 2025 at 4:36 PM Willem Kaufmann ***@***.***>
wrote:

> *WillemKauf* left a comment (redpanda-data/redpanda#27025)
> <https://github.com/redpanda-data/redpanda/pull/27025#issuecomment-3144984081>
>
> That's a lot of context to keep in your head to reason about correctness.
> Why we don't throw if recovery is aborted? To me that looks like way more
> straightforward mechanism and then you don't have to worry at all about
> someone using a bogus return value inadvertently.
>
> Thanks, I do agree. I was trying to find a way to maintain status quo
> behavior, but I agree, segment_set recovery should instead behave this
> way.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/redpanda-data/redpanda/pull/27025#issuecomment-3144984081>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEETWIDPDBCEQJMPNCNSJ33LOCQXAVCNFSM6AAAAACCSR4WCGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCNBUHE4DIMBYGE>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

WillemKauf commented:
> One thing I wonder about is whether the previous behavior was there to support making incremental progress even if the server is shutdown.

There is a try/catch block much above this in `controller_backend`:

https://github.com/redpanda-data/redpanda/blob/98cf9ebcb8dadce94a7a0575212118ee59d19116/src/v/cluster/controller_backend.cc#L1482-L1505

so the reconciliation step will fail there, but shutdown should proceed as normal.

dotnwat commented:
Taking another look at this I'm not sure I understand the motivation. Originally there was this added check for abort requested inside `maybe_create_contiguous_segment_set`:

```
    for (auto it = std::next(segs.begin()); it != segs.end();) {
        if (as.abort_requested()) {
            co_return segment_set(std::move(segs));
        }
```

and from what I can tell the only I/O that function does is `co_await rename` on ignored segments. so is this change even necessary?

WillemKauf commented:
> so is this change even necessary?

No, I don't think so. I was investigating a hang that was ultimately fixed by https://github.com/redpanda-data/redpanda/pull/27050 so I'm happy to close this without merging.

dotnwat commented:
> > so is this change even necessary?
> 
> No, I don't think so. I was investigating a hang that was ultimately fixed by #27050 so I'm happy to close this without merging.

Ok, sgtm. But feel free to re-open if you think there is an important performance improvement here. I'm just not seeing it.
